src/driftpy/user_map/user_map.py

```diff
@@ -0,0 +1,156 @@
+    
+    def get_user_authority(self, key: str) -> Optional[Pubkey]:
+        ch_user = self.user_map.get(key)
+        if not ch_user:
+            return None
+        return ch_user.get_user_account().authority
+    
+    async def must_get(self, key: str) -> DriftUser:
+        if not self.has(key):
+            pubkey = Pubkey.from_string(key)
+            await self.add_pubkey(pubkey)
+        return self.get(key)
+    
+    async def add_pubkey(
+            self, 
+            user_account_public_key: Pubkey, 
+        ) -> None:
+        if isinstance(self.subscription, PollingSubscription):
+            bulk_account_loader = BulkAccountLoader(self.drift_client.connection)
+            config = AccountSubscriptionConfig(ConfigType.POLLING.value, bulk_account_loader, self.commitment)
+        elif isinstance(self.subscription, WebsocketSubscription):
+            config = AccountSubscriptionConfig(ConfigType.WEBSOCKET.value, commitment = self.commitment)
+        else:
+            config = AccountSubscriptionConfig(ConfigType.CACHED.value)
+        user = DriftUser(
+            self.drift_client, 
+            authority = user_account_public_key,
+            account_subscription = config
+            )
+        await user.subscribe()
+        self.user_map[str(user_account_public_key)] = user
+
+    async def sync(self):
+        async with self.sync_lock:
+            try:
+                filters = (get_user_filter(),)
+                if not self.include_idle:
+                    filters += (get_non_idle_user_filter(),)
+
+                rpc_json_response = await self.connection.get_program_accounts(self.drift_client.program_id, self.commitment, 'base64', filters=filters)
+                rpc_response_and_context = rpc_json_response.value
+
+                slot = (await self.drift_client.program.provider.connection.get_slot()).value
+                program_account_buffer_map = {}
+
+                for program_account in rpc_response_and_context:
+                    pubkey = program_account.pubkey
+                    data = self.drift_client.program.coder.accounts.decode(program_account.account.data)
+                    program_account_buffer_map[str(pubkey)] = data
+
+                for key in program_account_buffer_map.keys():
+                    if key not in self.user_map:
+                        data = program_account_buffer_map.get(key)
+                        user_account = self.drift_client.program.coder.accounts.decode(data)
+                        await self.add_pubkey(Pubkey.from_string(key))
+                    await asyncio.sleep(0)
+
+                for key, user in self.user_map.items():
+                    if key not in program_account_buffer_map:
+                        user.unsubscribe()
+                        del self.user_map[key]
+                    else:
+                        user_account = self.drift_client.program.coder.accounts.decode(program_account_buffer_map.get(key))
```

Comment:
recently edited how this worked in ts such that the update data call happens in for loop above, which more performant: https://github.com/drift-labs/protocol-v2/blob/master/sdk/src/userMap/userMap.ts#L335